proposals/0001-metadata-providers.md

```diff
@@ -0,0 +1,489 @@
+on `Logger.MetadataProvider`:
+
+```swift
+extension Logger.MetadataProvider {
+    static let tracer = Logger.MetadataProvider { baggage in
+        guard let spanContext = baggage?.spanContext else { return nil }
+        return [
+          "traceID": "\(spanContext.traceID)",
+          "spanID": "\(spanContext.spanID)",
+        ]
+    }
+}
+```
+
+### Using a `MetadataProvider`
+
+A `MetadataProvider` can be set up either globally, on a boot-strapped logging system:
+
+```swift
+LoggingSystem.bootstrap(
+    metadataProvider: .tracer,
+    StreamLogHandler.standardOutput
+)
+```
+
+or, on a specific logger, in case some contexts should extract additional information:
+
+```swift
+let logger = Logger(label: "example", metadataProvider: .init { baggage in
+    guard let operationID = baggage?.operationID else { return nil }
+    return ["extra/opID": "\(opID)"]
+})
+```
+
+> NOTE: Setting the metadata provider on the logger directly means the `LoggingSystem` metadata provider
+> is skipped (if defined), following how an explicitly passed handler `factory`
+> overrides the `LoggingSystem`s `factory`.
+
+Once a metadata provider was set up, when a log statement is about to be emitted, the logger will obtain the task-local `Baggage.current`,
+and pass it to the metadata providers. The providers can then convert specific baggage items into logger metadata as they see fit.
+
+Those metadata will then be included in the log statement, e.g. like this:
+
+```swift
+var baggage = Baggage.topLevel
+baggage.spanContext = SpanContext()
+Baggage.$current.withValue(baggage) {
+    test()
+}
+
+func test() {
+    log.info("Test", metadata: ["oneOff": "42"])
+    // info [traceID: abc, spanID: 123, onOff: 42] Test
+}
+```
+
+### Multiple `MetadataProvider`s using `MetadataProvider.multiplex(_:)`
+
+Borrowing the concept from log handlers, metadata providers also have a multiplexing implementation.
+It is defined as an extension on `MetadataProvider` and is useful in cases where users want to utilize
+more than one metadata provider at the same time:
+
+```swift
+extension Logger.MetadataProvider {
+    public static func multiplex(_ providers: [Logger.MetadataProvider]) -> Logger.MetadataProvider {
+        assert(!providers.isEmpty, "providers MUST NOT be empty")
+        return Logger.MetadataProvider { baggage in
+            providers.reduce(into: nil) { metadata, provider in
+                if let providedMetadata = provider.metadata(baggage) {
+                    if metadata != nil {
+                        metadata!.merge(providedMetadata, uniquingKeysWith: { _, rhs in rhs })
+                    } else {
+                        metadata = providedMetadata
+                    }
+                }
+            }
+        }
+    }
+}
+```
+
+Metadata keys are unique, so in case multiple metadata providers return the same key,
+the last provider in the array "wins" and provides the value.
```

Comment:
Should we do this silently here? I think some of our `LogHandler`s actually precondition or fatalError on duplicates to avoid silent duplicates that override something.